aten/src/ATen/native/SobolEngineOps.cpp

```diff
@@ -0,0 +1,144 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+#include <ATen/native/SobolEngineOpsUtils.h>
+
+#include <vector>
+
+namespace at {
+namespace native {
+
+/// This is the core function to draw samples from a `SobolEngine` given
+/// its state variables (`sobolstate` and `quasi`). `dimension` can be
+/// inferred from `sobolstate`, but choosing to pass it explicitly to avoid
+/// an extra operation to obtain the size of the first dimension of
+/// `sobolstate`.
+std::tuple<Tensor, Tensor> _sobol_engine_draw(const Tensor& quasi, int64_t n, const Tensor& sobolstate,
+                                              int64_t dimension, int64_t num_generated) {
+  AT_CHECK(sobolstate.dtype() == at::kLong,
+           "sobolstate needs to be of type ", at::kLong);
+  AT_CHECK(quasi.dtype() == at::kLong,
+           "quasi needs to be of type ", at::kLong);
+
+  /// Performing one `unbind` operation and caching the result to prevent `n`
+  /// `select` operations.
+  std::vector<Tensor> sobolstate_unbind = at::native::unbind(sobolstate, 1);
+
+  /// Considering a vector of `n` Tensors to store the results in.
+  std::vector<Tensor> result(n);
+  Tensor wquasi = quasi.clone();
+
+  int64_t l;
+  for (int64_t i = 0; i < n; ++i) {
+    l = rightmost_zero(num_generated);
+    result[i] = wquasi.__ixor__(sobolstate_unbind[l]).clone();
```

Comment:
Why are you `clone`ing the tensor? Shouldn't `wquasi.__ixor__(sobolstate_unbind[l])` return a new tensor?

Reply:
@zou3519 `__ixor__` is an inplace method, and hence returns a reference. The lack of clone seemed to be erroneous, as the autograd engine was complaining something about one of `copy_` or `copy_from` being non-differentiable.

This issue doesn’t exist in the patch above 
https://github.com/pytorch/pytorch/pull/10505#issuecomment-474427729

Reply:
Oh, I see. It will probably be slightly faster to do `wquasi.__xor__(sobolstate_unbind[l]).clone()` (I think `__xor__` is the out-of-place variant of `__ixor__`?) so that it only needs to make one library call

Reply:
The reason we consider the inplace version is because the value of `wquasi` is intentionally modified every iteration.